### ISSUE TITLE
Implement Embeddings System with BGE-small via MLX Swift

### ISSUE THREAD

**Issue by AtlantisPleb** [OpenAgentsInc/openagents#1467]
# Embeddings System Implementation

## Overview

Implement a reusable embeddings primitive for OpenAgents using the **BGE-small-en-v1.5** model from Hugging Face via MLX Swift. The macOS desktop app will download and run the model locally, providing embedding generation and semantic search capabilities to all clients (iOS app, agents, orchestration) via the existing JSON-RPC bridge.

**Related Documentation:**
- 📋 [Detailed Implementation Plan](../docs/plans/embeddings-implementation-plan.md)
- 📚 [Embeddings Research](../docs/research/20251110-embeddings.md)
- 🔍 [SearchKit Spec v0.2.2](../docs/engine/spec-v0.2.2.md)

## Key Design Principles

1. **Reusable & Swappable**: Abstract interface allows model switching (local MLX, cloud APIs, Apple frameworks)
2. **macOS-Hosted**: All model loading and inference on macOS; iOS/agents consume via bridge
3. **Actor-Based**: Thread-safe service following existing patterns (`SessionUpdateHub`, `TinyvexDbLayer`)
4. **Storage-Ready**: SQLite vector storage with L2-normalized Float32 arrays
5. **ACP-Aligned**: JSON-RPC methods follow ACP conventions (snake_case, typed responses)

## Primary Use Cases

- **SearchKit**: Semantic code search, hybrid retrieval (FTS5 + embeddings)
- **Orchestration**: Context-aware file selection for agent exploration
- **Clustering**: Group similar conversations, files, or code blocks
- **RAG**: Retrieval-augmented generation for on-device LLM
- **Future**: Duplicate detection, recommendation, anomaly detection

### New Module Structure

\`\`\`
ios/OpenAgentsCore/Sources/OpenAgentsCore/
├── Embeddings/                          # NEW MODULE
│   ├── EmbeddingProvider.swift          # Protocol for swappable providers
│   ├── MLXEmbeddingProvider.swift       # MLX-based implementation
│   ├── EmbeddingService.swift           # Main service actor (macOS-only)
│   ├── VectorStore.swift                # SQLite vector storage actor
│   ├── EmbeddingTypes.swift             # Request/response types
│   └── SimilaritySearch.swift           # Cosine search + utilities
\`\`\`

### Bridge Integration

- JSON-RPC methods: `embedding/generate`, `embedding/store`, `embedding/search`, `embedding/store_batch`
- Local RPC mappings for macOS app direct access
- Extends existing `DesktopWebSocketServer` patterns

### Storage Extension

- New SQLite table: `embeddings` (id, collection, embedding_blob, dimensions, model_id, metadata, timestamps)
- Float32 vectors, L2-normalized for cosine similarity via dot product
- Brute-force search (Phase 1-5), ANN indexing (Phase 6+)

## Model Selection

**Primary Model:** `mlx-community/bge-small-en-v1.5-6bit`

**Characteristics:**
- Dimensions: 384
- Size: ~45 MB (6-bit quantized)
- Quality: Strong retrieval performance (MTEB benchmark)
- Speed: Fast on Apple Silicon (M1+)
- Cache: `~/.cache/huggingface/hub/` (managed by MLX)

**Storage Footprint:**
- 384 dims × 4 bytes = 1,536 bytes per vector
- 100,000 vectors ≈ 146 MB

### Phase 1: Foundation (Week 1-2)
**Goal:** Core types, provider protocol, MLX implementation

- [ ] Add MLX Swift Examples dependency to Package.swift
- [ ] Implement `EmbeddingTypes.swift` (request/response types, errors)
- [ ] Implement `EmbeddingProvider.swift` (protocol)
- [ ] Implement `MLXEmbeddingProvider.swift` (BGE-small via MLX)
- [ ] Write unit tests (`EmbeddingsTests.swift`)

**Acceptance Criteria:**
- Can load BGE-small model
- Can generate normalized embeddings
- Unit tests pass on macOS

### Phase 2: Storage (Week 2-3)
**Goal:** SQLite vector storage and search

- [ ] Extend `TinyvexDbLayer` with `embeddings` table schema
- [ ] Implement `VectorStore.swift` (store, fetch, search)
- [ ] Implement cosine similarity search (Accelerate framework)
- [ ] Write integration tests (`EmbeddingServiceIntegrationTests.swift`)

**Acceptance Criteria:**
- Can store/retrieve embeddings
- Search returns correct results by similarity
- Tests cover storage round-trip

### Phase 3: Service & Bridge (Week 3-4)
**Goal:** Service coordination and bridge exposure

- [ ] Implement `EmbeddingService.swift` (main actor coordinating provider + storage)
- [ ] Add `DesktopWebSocketServer+Embeddings.swift` (JSON-RPC methods)
- [ ] Extend `DesktopWebSocketServer+Local.swift` (local mappings)
- [ ] Initialize service in `DesktopWebSocketServer.init()`
- [ ] Write bridge tests (`EmbeddingBridgeTests.swift`)

**Acceptance Criteria:**
- macOS app can generate embeddings via local API
- JSON-RPC methods work over WebSocket
- iOS app can call via bridge

### Phase 4: SearchKit Integration (Week 4-6)
**Goal:** Hybrid search (FTS5 + semantic) for codebase exploration

- [ ] Implement `SearchKitService.swift` (uses `EmbeddingService`)
- [ ] Implement hybrid search (Reciprocal Rank Fusion)
- [ ] Add file indexing with embeddings
- [ ] Integrate with `ExploreOrchestrator`
- [ ] Add to Foundation Models tools (`FMTools.swift`)
- [ ] Write SearchKit E2E tests

**Acceptance Criteria:**
- Can index OpenAgents repo
- Hybrid search returns relevant files
- Orchestrator uses semantic search for file selection

### Phase 5: Optimization & Polish (Week 6-7)
**Goal:** Performance, error handling, documentation

- [ ] Write performance benchmarks (`EmbeddingPerformanceTests.swift`)
- [ ] Optimize batch embedding (tune batch size)
- [ ] Memory profiling and optimization
- [ ] Polish error handling and messages
- [ ] Write API documentation (DocC)
- [ ] Create ADR-0008 (Embeddings Architecture)

**Acceptance Criteria:**
- Single embedding <50ms p50
- Batch 100 embeddings <2s
- Search 10k vectors <500ms
- All tests pass

### Phase 6: Provider Extensibility (Week 7-8, Optional)
**Goal:** Swappable providers and configuration

- [ ] Implement `AppleNLEmbeddingProvider.swift` (NaturalLanguage fallback)
- [ ] Implement `CloudEmbeddingProvider.swift` (OpenAI/Anthropic)
- [ ] Add Settings UI for provider selection (macOS)
- [ ] Implement provider swapping logic
- [ ] Write provider tests

**Acceptance Criteria:**
- Can switch providers without code changes
- Fallback to NaturalLanguage if MLX unavailable
- Cloud provider respects policy gates

### New SwiftPM Dependency

\`\`\`swift
// In ios/OpenAgentsCore/Package.swift
dependencies: [
    .package(url: "https://github.com/ml-explore/mlx-swift-examples.git", 
             from: "0.18.0"),
],
targets: [
    .target(
        name: "OpenAgentsCore",
        dependencies: [
            .product(name: "MLXEmbedders", package: "mlx-swift-examples"),
        ]
    ),
]
\`\`\`

### System Requirements

**macOS:**
- macOS 13.0+ (existing requirement)
- Apple Silicon (M1+) or Intel with GPU (for MLX acceleration)
- ~200 MB disk space (model + dependencies)
- ~500 MB RAM (model loaded)

**iOS:**
- iOS 16.0+ (existing requirement)
- No embedding computation (bridge client only)

### Performance
- **Latency:** <50ms p50, <100ms p95 (single embedding)
- **Throughput:** >50 embeddings/sec (batch mode)
- **Search:** <500ms p95 for 10k vectors (brute-force)
- **Memory:** <500 MB RAM for model + index

### Quality
- **Relevance:** SearchKit hybrid search improves over FTS5-only
- **Recall:** Semantic search finds relevant files missed by lexical search
- **Precision:** Top-5 results have >80% relevance

### Reliability
- **Availability:** Model loads successfully >99% of launches
- **Errors:** <1% embedding generation failures (excl. network)
- **Crash-free:** No crashes attributed to embedding service

### Phase 7+: Advanced Features
- **ANN Indexing** (HNSW, sqlite-vec) for >100k vectors
- **Reranking** (Foundation Models or small reranker model)
- **Multilingual Models** (Arctic Embed, Qwen3)
- **Contextual Code Embeddings** (with imports/context)
- **On-Device Fine-Tuning** (MLX training APIs)

## Testing Strategy

- **Unit Tests:** Provider, normalization, similarity computation
- **Integration Tests:** Service, storage, round-trip
- **Bridge Tests:** JSON-RPC methods, local mappings
- **Performance Tests:** Latency, throughput, search speed
- **E2E Tests:** SearchKit hybrid search, orchestration integration

## Documentation

- [x] Implementation Plan (`docs/plans/embeddings-implementation-plan.md`)
- [ ] ADR-0008 (Embeddings Architecture)
- [ ] User Documentation (`docs/embeddings/overview.md`, `api.md`, `troubleshooting.md`)
- [ ] Developer Guide (`docs/embeddings/developer-guide.md`)
- [ ] API Reference (DocC inline comments)

## Open Questions

1. **Model Updates:** Pin specific version or allow auto-update?
   - Recommendation: Pin for stability, manual update in Phase 6
   
2. **Storage Format:** Float32 vs Float16?
   - Recommendation: Start with Float32, migrate to Float16 if storage becomes issue
   
3. **Batch Size:** What size for optimal throughput?
   - Recommendation: Start with 16, make configurable in Phase 5
   
4. **Collection Strategy:** How to organize collections?
   - Recommendation: Hybrid (by type + workspace metadata filtering)

## Risk Mitigation

- **Model Download Failure:** Graceful degradation to FTS5-only, retry logic, clear errors
- **Memory Pressure:** Monitor usage, unload model when idle, paginated search
- **Performance Regression:** Background embedding, batching, caching, performance tests in CI
- **Model Compatibility:** Pin MLX version, comprehensive tests, monitor release notes

---

**Total Estimated Effort:** 6-8 weeks (5 core phases + 1 optional extensibility phase)

**Next Steps:**
1. Review and approve implementation plan
2. Begin Phase 1 (Foundation)
3. Iterate based on learnings and feedback

See [full implementation plan](../docs/plans/embeddings-implementation-plan.md) for detailed code examples, API specifications, and technical details.

**Comment by AtlantisPleb** [comment]
Status update (2025-11-10 13:06 PT)

Embeddings Phase 1–3 implemented and bridged:

- Provider/Service/Store:
  - MLXEmbeddingProvider, VectorStore (Accelerate cosine), EmbeddingService actors
- DB + RPC:
  - Tinyvex embeddings schema + helpers; JSON-RPC: generate/store/search/store_batch
  - Desktop bridge handlers in DesktopWebSocketServer+Embeddings.swift
- Clients:
  - SearchKit docs updated to depend on EmbeddingService
  - SQLite cosine options documented (sqlite-vec/libSQL) for Phase 2+

Code pointers:
- ios/OpenAgentsCore/Sources/OpenAgentsCore/Embeddings/
- ios/OpenAgentsCore/Sources/OpenAgentsCore/Tinyvex/DbLayer.swift
- ios/OpenAgentsCore/Sources/OpenAgentsCore/DesktopBridge/DesktopWebSocketServer+Embeddings.swift
- docs/engine/sqlite-cosine-vector-search.md

Build: macOS target builds clean via xcodebuild; tests pending.

Next:
- Wire hybrid search (FTS5 + semantic + RRF)
- Add SearchKit→EmbeddingService integration tests
- Consider sqlite-vec adapter for large corpora
